### PR TITLE
[consul] Add instrumentation for consul

### PR DESCRIPTION
**Pull request by phil-dd**
Automatically instruments consul connections to trace KV calls. Only works for sync client.

### CODE REVIEW

File: ddtrace/__init__.py [hunk 1]
```diff
@@ -11,7 +11,7 @@
     __version__ = pkg_resources.get_distribution(__name__).version
 except pkg_resources.DistributionNotFound:
     # package is not installed
-    __version__ = None
+    __version__ = 'not_installed'
```
Comment (brettlangdon) [inline ddtrace/__init__.py]: Why this change?

File: ddtrace/contrib/consul/__init__.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+"""
+
+from ...utils.importlib import require_modules
+
+required_modules = ['consul']
+
+with require_modules(required_modules) as missing_modules:
+    if not missing_modules:
+        from .patch import patch
+        __all__ = ['patch']
```
Comment (brettlangdon) [inline ddtrace/contrib/consul/__init__.py]: can we export `unpatch` as well?

File: ddtrace/contrib/consul/patch.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        if not isinstance(instance.agent.http, consul.std.HTTPClient):
+            return original_func(*args, **kwargs)
+
+        path = args[0]
+
+        with pin.tracer.trace(name, service=pin.service, resource=path) as span:
+            span.set_tag('consul.key', path)
+            return original_func(*args, **kwargs)
+
+    return wrapping_func
```
Comment (brettlangdon) [inline ddtrace/contrib/consul/patch.py]: missing newline at end of file

File: ddtrace/contrib/consul/patch.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    def wrapping_func(original_func, instance, args, kwargs):
+        pin = Pin.get_from(instance)
+        if not pin or not pin.enabled():
+            return original_func(*args, **kwargs)
+
+        # Only patch the syncronous implementation
+        if not isinstance(instance.agent.http, consul.std.HTTPClient):
+            return original_func(*args, **kwargs)
+
+        path = args[0]
```
Comment (brettlangdon) [inline ddtrace/contrib/consul/patch.py]: could `path` ever be passed as a `kwargs`?

in others we do:

```python
path = kwargs.get('path', args[0])
```

(ish)

File: ddtrace/__init__.py (hunk 1 above)
Comment (phil-dd) [reply ddtrace/__init__.py]: Setting `__version__` to `None` results in an error when it attempts to set headers, since the value of a header can't be None.

File: ddtrace/contrib/consul/__init__.py (hunk 2 above)
Comment (phil-dd) [reply ddtrace/contrib/consul/__init__.py]: Done

File: ddtrace/__init__.py (hunk 1 above)
Comment (brettlangdon) [reply ddtrace/__init__.py]: how are you using it? this hasn't been an issue for any of us before.

can you remove from this PR, if we make this change it should be in it's own PR

File: ddtrace/contrib/consul/patch.py (hunk 3 above)
Comment (phil-dd) [reply ddtrace/contrib/consul/patch.py]: Fixed

File: ddtrace/__init__.py (hunk 1 above)
Comment (phil-dd) [reply ddtrace/__init__.py]: I cloned the repo and symlinked `ddtrace` into a virtualenv for testing. I'll remove the change.

File: ddtrace/contrib/consul/patch.py (hunk 4 above)
Comment (phil-dd) [reply ddtrace/contrib/consul/patch.py]: Changed to try getting from kwargs, and added a test to ensure it works

File: ddtrace/__init__.py (hunk 1 above)
Comment (brettlangdon) [reply ddtrace/__init__.py]: ah, should do `python setup.py develop` for that

File: ddtrace/contrib/consul/patch.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        return
+    setattr(consul, '__datadog_patch', False)
+
+    for f_name in _KV_FUNCS:
+        name = f_name.split('.')[-1]
+        _u(consul.Consul.KV, name)
+
+
+def wrap_function(name):
+    def wrapping_func(original_func, instance, args, kwargs):
```
Comment (majorgreys) [inline ddtrace/contrib/consul/patch.py]:
```suggestion
    def trace_func(wrapped, instance, args, kwargs):
```

Minor suggestion for function name and argument name to keep consistent with codebase.

File: ddtrace/contrib/consul/patch.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+
+def patch():
+    if getattr(consul, '__datadog_patch', False):
+        return
+    setattr(consul, '__datadog_patch', True)
+
+    pin = Pin(service='consul', app='consul', app_type=AppTypes.cache)
+    pin.onto(consul.Consul)
+    pin.onto(consul.Consul.KV)
```
Comment (majorgreys) [inline ddtrace/contrib/consul/patch.py]: How are the multiple pins here intended to be used? Are both needed for the inheriting the pin used in `wrap_function`?

Comment (phil-dd) [reply ddtrace/contrib/consul/patch.py]: The pin to `consul.Consul` could potentially be removed - this was done because [the consul lib implements the KV as a child class of Consul](https://github.com/cablehead/python-consul/blob/53eb41c4760b983aec878ef73e72c11e0af501bb/consul/base.py#L462) so I wasn't sure what correct practice is here.
